src/Shared/FileUtilities.cs

```diff
@@ -254,165 +256,97 @@ internal static String GetDirectoryNameOfFullPath(String fullPath)
-
-                            /*
-                              From Path.cs in the CLR
-
-                              Throw an ArgumentException for paths like \\, \\server, \\server\
-                              This check can only be properly done after normalizing, so
-                              \\foo\.. will be properly rejected.  Also, reject \\?\GLOBALROOT\
-                              (an internal kernel path) because it provides aliases for drives.
-
-                              throw new ArgumentException(Environment.GetResourceString("Arg_PathIllegalUNC"));
-
-                               // Check for \\?\Globalroot, an internal mechanism to the kernel
-                               // that provides aliases for drives and other undocumented stuff.
-                               // The kernel team won't even describe the full set of what
-                               // is available here - we don't want managed apps mucking
-                               // with this for security reasons.
-                            */
-                    if (startIndex == finalFullPath.Length || finalFullPath.IndexOf(@"\\?\globalroot", PathComparison) != -1)
-                    {
-                        finalFullPath = Path.GetFullPath(finalFullPath);
-                    }
-                }
+        private static string GetFullPath(string path)
+        {
+#if FEATURE_LEGACY_GETFULLPATH
+            if (NativeMethodsShared.IsWindows)
+            {
+                string uncheckedFullPath = NativeMethodsShared.GetFullPath(path);
+                AssertMaxPathLimits(uncheckedFullPath);
 
-                        return finalFullPath;
-                    }
-                }
+                // We really don't care about extensions here, but Path.HasExtension provides a great way to
+                // invoke the CLR's invalid path checks (these are independent of path length)
+                Path.HasExtension(uncheckedFullPath);
 
-                NativeMethodsShared.ThrowExceptionForErrorCode(errorCode);
-                return null;
+                // If we detect we are a UNC path then we need to use the regular get full path in order to do the correct checks for UNC formatting
+                // and security checks for strings like \\?\GlobalRoot
+                return IsUNCPath(uncheckedFullPath) ? Path.GetFullPath(uncheckedFullPath) : uncheckedFullPath;
             }
 #endif
-            return FixFilePath(Path.GetFullPath(path));
-
+            string fullPath = Path.GetFullPath(path);
+            AssertMaxPathLimits(fullPath);
+            return fullPath;
         }
 
-        internal static string NormalizePath(string directory, string file)
+        private static void AssertMaxPathLimits(string path)
         {
-            return NormalizePath(Path.Combine(directory, file));
+            if (IsPathTooLong(path))
+            {
+                throw new System.IO.PathTooLongException();
```

Comment:
Need to supply a message. If it's just the path to the file that's easy, otherwise you'll need to add a string resource saying the path is too long and add the path to the file.